internal/buildsign/ocpbuild/ocpbuildmanager.go

```diff
@@ -0,0 +1,173 @@
+package ocpbuild
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	buildv1 "github.com/openshift/api/build/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/rh-ecosystem-edge/kernel-module-management/internal/constants"
+)
+
+type Status string
+
+const (
+	ocpbuildTypeBuild = "build"
+	ocpbuildTypeSign  = "sign"
+
+	StatusCompleted  Status = "completed"
+	StatusCreated    Status = "created"
+	StatusInProgress Status = "in progress"
+	StatusFailed     Status = "failed"
+
+	hashAnnotation = "kmm.node.kubernetes.io/last-hash"
+)
+
+var ErrNoMatchingBuild = errors.New("no matching Build")
+
+//go:generate mockgen -source=ocpbuildmanager.go -package=ocpbuild -destination=mock_ocbuildmanager.go
+
+type ocpbuildManager interface {
+	isOCPBuildChanged(existingBuild *buildv1.Build, newBuild *buildv1.Build) (bool, error)
+	getModuleOCPBuildByKernel(ctx context.Context, modName, namespace, kernelVersion, ocbuildType string, owner metav1.Object) (*buildv1.Build, error)
+	getModuleOCPBuilds(ctx context.Context, modName, modNamespace, ocbuildType string, owner metav1.Object) ([]buildv1.Build, error)
+	getOCPBuildStatus(build *buildv1.Build) (Status, error)
+	createOCPBuild(ctx context.Context, build *buildv1.Build) error
+	deleteOCPBuild(ctx context.Context, build *buildv1.Build) error
+	ocpbuildLabels(modName, kernelVersion, ocpbuildType string) map[string]string
+}
+
+type ocpbuildManagerImpl struct {
+	client client.Client
+}
+
+func newOCPBuildManager(client client.Client) ocpbuildManager {
+	return &ocpbuildManagerImpl{
+		client: client,
+	}
+}
+
+func (omi *ocpbuildManagerImpl) isOCPBuildChanged(existingBuild *buildv1.Build, newBuild *buildv1.Build) (bool, error) {
+	existingAnnotations := existingBuild.GetAnnotations()
+	newAnnotations := newBuild.GetAnnotations()
+	if existingAnnotations == nil {
+		return false, fmt.Errorf("annotations are not present in the existing build %s", existingBuild.Name)
+	}
+	if existingAnnotations[hashAnnotation] == newAnnotations[hashAnnotation] {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (omi *ocpbuildManagerImpl) getModuleOCPBuildByKernel(ctx context.Context,
+	modName,
+	modNamespace,
+	kernelVersion,
+	ocbuildType string,
+	owner metav1.Object) (*buildv1.Build, error) {
+	buildList := buildv1.BuildList{}
+
+	opts := []client.ListOption{
+		client.MatchingLabels(omi.ocpbuildLabels(modName, kernelVersion, ocbuildType)),
+		client.InNamespace(modNamespace),
+	}
+
+	if err := omi.client.List(ctx, &buildList, opts...); err != nil {
+		return nil, fmt.Errorf("could not list Build: %v", err)
+	}
+
+	// filter OCP builds by owner, since they could have been created by the preflight
+	// when checking that specific module
+	moduleOwnedOCPBuilds := filterOCPBuildsByOwner(buildList.Items, owner)
+
+	if n := len(moduleOwnedOCPBuilds); n == 0 {
+		return nil, ErrNoMatchingBuild
```

Comment:
Why is it an error? Can we just return nil with no error if no builds exist for the module?

Reply:
we want a specific error that says that the Build does not exists, but that the operation itself have not failed. The same as client returns when you try to get an object and the operation is successful but the object was not found